pandas/tseries/period.py

```diff
@@ -271,9 +271,14 @@ def _from_arraylike(cls, data, freq, tz):
 
     @classmethod
     def _simple_new(cls, values, name=None, freq=None, **kwargs):
-        if not getattr(values, 'dtype', None):
-            values = np.array(values, copy=False)
-        if is_object_dtype(values):
+
+        if lib.is_integer_array(np.array(values, copy=False)):
+            values = np.array(values, copy=False, dtype='int64')
+        elif (len(values) > 0 and
+                  lib.is_float_array(np.array(values, copy=False))):
+            # __new__ coerces to ints, so raise here
+            raise ValueError('Must supply ints or Period objects')
```

Comment:
I'd vote to always fail on floats, including in `__new__`, but I imagine there's some reason it coerces